core/src/main/java/org/elasticsearch/index/query/HasChildQueryBuilder.java

```diff
@@ -120,4 +177,242 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+
+            LateParsingQuery that = (LateParsingQuery) o;
+
+            if (minChildren != that.minChildren) return false;
+            if (maxChildren != that.maxChildren) return false;
+            if (!toQuery.equals(that.toQuery)) return false;
+            if (!innerQuery.equals(that.innerQuery)) return false;
+            if (!parentType.equals(that.parentType)) return false;
+            if (scoreMode != that.scoreMode) return false;
+            return parentChildIndexFieldData.equals(that.parentChildIndexFieldData);
+        }
+
+        @Override
+        public int hashCode() {
+            int result = super.hashCode();
+            result = 31 * result + toQuery.hashCode();
+            result = 31 * result + innerQuery.hashCode();
+            result = 31 * result + minChildren;
+            result = 31 * result + maxChildren;
+            result = 31 * result + parentType.hashCode();
+            result = 31 * result + scoreMode.hashCode();
+            result = 31 * result + parentChildIndexFieldData.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString(String s) {
+            return "LateParsingQuery {parentType=" + parentType + "}";
+        }
+
+        public int getMinChildren() {
+            return minChildren;
+        }
+
+        public int getMaxChildren() {
+            return maxChildren;
+        }
+
+        public ScoreMode getScoreMode() {
+            return scoreMode;
+        }
+    }
+
+    @Override
+    protected boolean doEquals(HasChildQueryBuilder that) {
+        return Objects.equals(queryBuilder, that.queryBuilder)
+                && Objects.equals(childType, that.childType)
+                && Objects.equals(scoreType, that.scoreType)
+                && Objects.equals(minChildren, that.minChildren)
+                && Objects.equals(maxChildren, that.maxChildren)
+                && Objects.equals(shortCircuitCutoff, that.shortCircuitCutoff)
+                && Objects.equals(queryInnerHits, that.queryInnerHits);
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(queryBuilder, childType, scoreType, minChildren, maxChildren, shortCircuitCutoff, queryInnerHits);
+    }
+
+    protected HasChildQueryBuilder(StreamInput in) throws IOException {
+        childType = in.readString();
+        minChildren = in.readInt();
+        maxChildren = in.readInt();
+        shortCircuitCutoff = in.readInt();
+        final int ordinal = in.readVInt();
+        scoreType = ScoreType.values()[ordinal];
+        queryBuilder = in.readQuery();
+        if (in.readBoolean()) {
+            queryInnerHits = new QueryInnerHits(in);
+        }
+    }
+
+    @Override
+    protected HasChildQueryBuilder doReadFrom(StreamInput in) throws IOException {
+        return new HasChildQueryBuilder(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(childType);
```

Comment:
should it be writeOptionalString instead? I don't think so given that we will validate before sending over the wire (see comment above).